stock_picking_invoicing/models/stock.py

```diff
@@ -0,0 +1,458 @@
+            'invoice_line_tax_ids': [(6, 0, taxes_ids)],
+            'discount': 0.0,
+            'account_analytic_id': False,
+        }
+
+        #https://github.com/OCA/account-invoicing/blob/8.0/stock_picking_invoicing_unified/models/stock_move.py
+
+
+        # negative value on quantity
+        if ((inv_type == 'out_invoice' and
+            self.location_id.usage == 'customer') or
+            (inv_type == 'out_refund' and
+            self.location_dest_id.usage == 'customer') or
+            (inv_type == 'in_invoice' and
+            self.location_dest_id.usage == 'supplier') or
+            (inv_type == 'in_refund' and
+            self.location_id.usage == 'supplier')):
+            res['quantity'] *= -1
+
+        return res
+        
+        
+    @api.multi
+    def _get_price_unit_invoice(self, inv_type):
+        """ Gets price unit for invoice
+        @param move_line: Stock move lines
+        @param type: Type of invoice
+        @return: The price unit for the move line
+        """
+        
+        if type in ('in_invoice', 'in_refund'):
+            return move_line.product_id.price        
+        else:
+            # If partner given, search price in its sale pricelist
+            if self.partner_id and self.partner_id.property_product_pricelist:
+                self = self.with_context(
+                                         partner=self.partner_id.id,
+                                         quantity=self.product_uom_qty,
+                                         date=self.date,
+                                         pricelist=self.partner_id.property_product_pricelist.id,
+                                         uom=self.product_uom.id
+                                            )
+        result = self.product_id.price        
+
+        return result
+    
+    @api.multi
+    def _get_moves_taxes(self, moves, inv_type):
+        
+        #extra moves with the same picking_id and product_id of a move have the same taxes
+        extra_move_tax = {}
+        is_extra_move = {}
+        for move in moves:
+            if move.picking_id:
+                is_extra_move[move.id] = True
+                if not (move.picking_id, move.product_id) in extra_move_tax:
+                    extra_move_tax[move.picking_id, move.product_id] = 0
+            else:
+                is_extra_move[move.id] = False
+        return (is_extra_move, extra_move_tax)
+    
+
+#----------------------------------------------------------
+# Picking
+#----------------------------------------------------------
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    #===========================
+    # Views
+    #===========================
+    
+    @api.multi
+    def set_to_be_invoiced(self):
+        for picking in self:
+            if picking.invoice_state == '2binvoiced':
+                raise Warning(_("Can't update invoice control for picking %s: "
```

Comment:
Invoice state cannot be updated for picking %s as it is already set as "To be invoiced"